### ISSUE TITLE
Hypernetwork API

### ISSUE THREAD

**Issue by jessebett** [FluxML/Flux.jl#797]
A [HyperNetwork](https://arxiv.org/pdf/1609.09106.pdf) is a neural network which outputs the parameters for another neural network. This issue is to discuss the implementation of a convenient API for expressing these in Flux.

The immediate difficulty working with hypernets is that Flux model creation prefers implicit dependence on parameters. This came up with neural ODEs and DiffEqFlux has a solution to make an implicitly parameterized function into an explicit one by `destructure`ing and `restructructure`ing. Issue #742 discusses bringing this, or some other solution to implicit->explicit, into Flux. 

Using the `DiffEqFlux: destructure, restructure` solution allows me to define a hypernet by first defining a model in the usual way, `destructure`ing to a vector of model parameters just to get the length. This gives the output dimension for a hypernet which is defined in the usual way as well.

```
hyper_dim = 20
nn_dim = 20

m_nn = Chain(Dense(1,nn_dim,tanh),Dense(nn_dim,1))
num_p = length(DiffEqFlux.destructure(m_nn))
hyper = Chain(Dense(1,hyper_dim,tanh),Dense(hyper_dim,hyper_dim,tanh),Dense(hyper_dim,num_p))

function model(x,t)
  p = hyper([t])
  m = restructure(m_nn,p)
  return m(x)
end
```
In this example the hypernet takes a single scalar hyperparameter, `t`. The `model` takes `x,t` , evaluates the hypernet on `t` to produce the parameters for the model neural net `m_nn` which is restructured and then evaluated on `x`.

@MikeInnes suggests we can have a much more elegant API for this, something like `m = Hypernet(100, Dense(10,5))` would take a 100-dim hyperparameter input and return a new `Dense(10,5)`.
The goal of this sort of API would be to have `Hypernet` be a layer in the chain which outputs a parameterized layer, e.g. `Dense`, or a chain of layers, e.g. the parameters to `m_nn` above. 
Then the model would look like `model(x,t) = hyper([t])(x)`

@MikeInnes in your proposed API you're having `Hypernet` return a layer: `Hypernet(100, Dense(10,5))` Do you anticipate difficulties having the hypernet return anything treelike? `Hypernet(100, Chain(Dense(10,5),Dense(5,10)))`?

**Comment by MikeInnes** [comment]
So `Hypernet` would just be:

```julia
Hypernet(n, m) =
  Chain(Dense(n, length(destructure(m))), w -> restructure(m, w))
```

Then your model is

```julia
hyper = Chain(Dense(1, 20, tanh), Dense(20, 20, tanh), Hypernet(20, m_nn))
model(x,t) = hyper([t])(x)
```

And that'll work for anything `treelike` (it's only a slight reordering of what @jessebett wrote anyway). We're technically slightly abusing the notion of a model here since the weights of `m_nn` never get used, but separating out model specifications seems like it'd be pretty inconvenient.

We probably want to bikeshed the vague names of `destructure` and `restructure` here (maybe something like `catparams`?) but beyond that I think this would be a very easy PR to accept.

**Comment by jessebett** [comment]
Autograd has a very useful convenience function we might want here, `flatten` on some nested structure, like model parameters, returns a flattened vector like `destructure`, as well as a function which `restructure`s those vectors back into the structure.
`p, restructure = flatten(m_nn)` then `restructure(p2)` could return the same structure as `m_nn` with the values changed.

I don't know whether this should overwrite the values inside `m_nn`, probably should not produce a new structure. 

A flatten like this could be written to dispatch on whatever kind of structure we want to store parameters in, dicts, named tuples, structs like `Dense`, and facilitate passing parameters between them through a vector. 

Then the hypernetwork would look like 
```
function hypernet(n,m)
  p ,restructure = flatten(m_nn)
  return Chain(Dense(n,length(p)), w->restructure(w))
end
```

This is slightly different, but the benefit of something like this is that it removes `destructure` and `restructure` in favour of `flatten`. I don't know what the tradeoffs for this would be.

**Comment by MikeInnes** [comment]
I quite like that way of writing it. It's really just a syntax question, because you can easily write one in terms of the other, e.g.

```julia
restructure(m, p) = flatten(m)[2](p)
flatten(m) = destructure(m), p -> restructure(m, p)
```

So there isn't a deep tradeoff here. But it sort-of avoids the issue where `restructure` takes a full model and only treats it as a model spec; if you want to ignore the weights then that's explicit. (Of course the Hypernet function would still ignore the weights implicitly, but maybe we should just encourage explicit `flatten`ing so it's clear what's going on).

`flatten` isn't an ideal name because we already have that in julia, but at least we now only have one name to figure out.

**Comment by jessebett** [comment]
How about we keep `destructure` as my proposed `restructure`. It makes it clear that it operates on structures of params, as well as gives symmetry to the returned `restructure`. But yes, this is nothing deep. It also would solve #742.

Alternatively instead of structure the name could also have to do with `parameterize`, since the `restructure` function in this case would be reparameterizing the model, but I feel `param` things already show up too much.

**Comment by MikeInnes** [comment]
Yeah, don't get me wrong, whatever happens #742 is part of this -- we might have Hypernet as a high-level thing but the tools it uses itself should definitely be exposed.

I'm not sure what you mean by "How about we keep `destructure` as my proposed `restructure`"; I guess the proposal is that we expose the `flatten` construct but keep the name `destructure` for it? So `p, restructure = destructure(m)`.

**Comment by jessebett** [comment]
Yes, that is what I'm proposing. Sorry for the word salad. `destructure` in lieu of `flatten` for the name.

**Comment by jessebett** [comment]
Actually, I don't think `Hypernet` should take in dimensions of input and the model it is parameterizing. I think it should look like `Hypernet(h,m)` where `h` is the hypernet model which outputs parameters for the `m` model. Then calling `hyper = Hypernet(h,m)` on an input `hyper(t)` produces a parameterized `m`.

@MikeInnes is the correct thing to do here to have the hypernet overwrite the parameter values in `m`? It shouldn't allocate a new chain every time it is called.

**Comment by jessebett** [comment]
To be clear:

```
function destructure(m)
  ps = []
  mapleaves(m) do p
    p isa TrackedArray && push!(ps, p)
    return p
  end
  return vcat(vec.(ps)...), ps->restructure(m,ps)
end

function restructure(m, ps)
  i = 0
  mapleaves(m) do p
    p isa TrackedArray || return p
    p = reshape(ps[i.+(1:length(p))], size(p))
    i += length(p)
    return p
  end
end
```

Can be used like
```
m = Dense(1,2)
ps, re = destructure(m)
m2=re(ps.*0) #zero the params and restructure them
destructure(m)[1] # unchanged
destructure(m2)[1] #zero'd params
```
My question is if this is correct, or should it be more like `restructure!` which mutates `m`?

**Comment by MikeInnes** [comment]
It's sort of appealing to make it mutating, but in general I think it's good to have the functional API first, and do the mutating version as an optimisation if needed.

It's probably a good idea to add a shape check to the `restructure` call, given that we know `length(p)` without any extra work.
